File: test/lua/framework/math/quadrature/sldfesq_test.lua
local function split_string(input, separator)
  local result = {}
  for value in input:gmatch("[^" .. separator .. "]+") do
    table.insert(result, value)
  end
  return result
end

local function getWeightSum(quad_points)
  log.Log(LOG_0, string.format("\nReading : %s", quad_points))
  local weight_sum = 0.0
  local file = io.open(quad_points, "r")
  if file then
    for line in file:lines() do
      local values = split_string(line, " ")
      local float_values = {}
      weight_sum = weight_sum + tonumber(values[4])
    end
    file:close()
  else
    log.Log(LOG_0ERROR, string.format("Error: Could not open file %s", quad_points))
  end
  return weight_sum
end

-- Qudrature-1 : Test creation of SLDFESQ with initial refinement level of 0
cquad1 = aquad.CreateSLDFESQuadrature(0)
aquad.PrintQuadratureToFile(cquad1, "TestQuad1_")
local quad1_sum = getWeightSum("TestQuad1_points.txt")
log.Log(LOG_0, string.format("Weight-Sum-1=%.3e\n\n", quad1_sum / (4 * math.pi)))

-- Qudrature-2 : Test local refinement of SLDFESQ with initial refinement level of 1
cquad2 = aquad.CreateSLDFESQuadrature(2)
aquad.LocallyRefineSLDFESQ(cquad2, { x = 0.25, y = -0.85, z = 1.0 }, 30.0 * math.pi / 180, false)
aquad.PrintQuadratureToFile(cquad2, "TestQuad2_")
local quad2_sum = getWeightSum("TestQuad2_points.txt")
log.Log(LOG_0, string.format("Weight-Sum-2=%.3e", quad2_sum / (4 * math.pi)))

os.execute("rm TestQuad1* TestQuad2*")

File: test/python/framework/math/quadrature/sldfesq_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
This script reads quadrature point files, sums a specific column of weights,
and prints normalized weight sums. It demonstrates creating a quadrature,
printing its points to a file, refining the quadrature locally, and cleaning up.
"""

import os
import math


def split_string(input_str, separator):
    """
    Splits the input_str using the specified separator.
    """
    # Using str.split() directly is sufficient for simple splits.
    return input_str.split(separator)


def get_weight_sum(quad_points_file):
    """
    Opens a quadrature points file, reads each line,
    splits the line to extract the 4th value (index 3) as a float,
    and sums these values.
    """
    print(f"\nReading: {quad_points_file}")
    weight_sum = 0.0
    try:
        with open(quad_points_file, "r") as file:
            file.readline()
            for line in file:
                # Remove any extra whitespace and split by space.
                # (Alternatively, you could simply use line.split() to split on any whitespace.)
                values = split_string(line.strip(), ",")
                weight_sum += float(values[3])
    except Exception:
        print(f"Error: Could not open file {quad_points_file}")
    print(weight_sum)
    return weight_sum


# --- Quadrature-1 ---
# Create a quadrature with initial refinement level of 0.
cquad1 = SLDFESQuadrature(level=0)
cquad1.PrintQuadratureToFile("TestQuad1")
quad1_sum = get_weight_sum("TestQuad1_points.csv")
print(f"Weight-Sum-1={(quad1_sum / (4 * math.pi)):.3e}\n\n")

# --- Quadrature-2 ---
cquad2 = SLDFESQuadrature(level=2)
# Locally refine the quadrature near a given point with a given angular spread.
cquad2.LocallyRefine(Vector3(0.25, -0.85, 1.0), 30.0 * math.pi / 180, False)
cquad2.PrintQuadratureToFile("TestQuad2")
quad2_sum = get_weight_sum("TestQuad2_points.csv")
print(f"Weight-Sum-2={(quad2_sum / (4 * math.pi)):.3e}\n\n")

# Clean up: remove generated files.
os.system("rm TestQuad1* TestQuad2*")
